Manual Docs 2

misteroneill opened the pull request:
Documentation updates to the tracks guides.

## Requirements Checklist
- [x] Feature implemented / Bug fixed
- [ ] Reviewed by Two Core Contributors

misteroneill committed:
tracks index

misteroneill committed:
Update tracks guides

brandonocasey commented on docs/guides/audio-tracks.md:
```diff
@@ -1,69 +1,80 @@
-Audio Tracks are a function of HTML5 video for providing alternative audio track selections to the user, so that a track other than the main track can be played. Video.js makes audio tracks work across all browsers. There are currently five types of tracks:
+## Caveats
+- It is not possible to add audio tracks through HTML. They must be added [programmatically](#api).
+- Literal switching of audio tracks for playback is not handled by Video.js and must be handled by something else. Video.js only stores the track representation.
```
maybe it would be a good idea to specify some examples here such as safari supporting them natively and HLS using them to support multiple audio tracks

brandonocasey commented on docs/guides/audio-tracks.md:
```diff
@@ -1,69 +1,80 @@
-### enabled
-If this track should be playing or not. In video.js we only allow one track to be enabled at a time. so if you enable more than one the last one to be enabled will end up being the only one.
+### Listen for a Track Becoming Enabled
+When a new track is enabled (other than the main track) an event is fired on the `AudioTrackList` called `"change"`. You can listen for that event and do something with it.
```
Maybe change this to "When a track is enable or disabled on an `AudioTrackList` a `change` event will be fired. You can listen for that event and do something with it.

> NOTE: The initial AudioTrack selection (usually the main track that is selected) should not fire a `change` event.

brandonocasey commented on docs/guides/audio-tracks.md:
```diff
@@ -1,69 +1,80 @@
+This class can be used to create new audio track objects. Each property below is available as an option to the `AudioTrack` constructor.
+
+#### `kind`
+One of the audio track types supported by Video.js:
```
maybe it would be good to specify that these are from the audiotrack spec?

brandonocasey commented on docs/guides/audio-tracks.md:
```diff
@@ -1,69 +1,80 @@
+The label for the track that will be shown to the user. For example, in a menu that lists the different languages available as alternate audio tracks.
+
+#### `language`
+The valid BCP 47 code for the language of the audio track, e.g. `"en"` for English or `"es"` for Spanish. For supported language translations, please see the [Languages Folder (/lang)](https://github.com/videojs/video.js/tree/master/lang) folder located in the Video.js root.
```
is there a guide for adding a language? if so maybe we should add a link to that guide here? if not maybe we should have one?

gkatsev replied:
link to BCP 47 explanation.

brandonocasey commented on docs/guides/audio-tracks.md:
```diff
@@ -1,69 +1,80 @@
-### language
-The two-letter code (valid BCP 47 language tag) for the language of the audio track, for example "en" for English. A list of language codes is [available here](languages.md#language-codes).
+// Add the track to the player's audio track list.
+player.audioTracks().addTrack(track);
```
It would be a good idea to add examples for the rest of the audioTrack and audioTrackList functions well. IE: `removeTrack`, `addTrack`, `getTrackById`

brandonocasey replied:
on second thought maybe we should have a general track guide/example so that it doesn't have to be repeated in every guide.

gkatsev replied:
It might be better to repeat these instead of having a generic tracks list.

misteroneill replied:
Yeah, the APIs are virtually identical, but the names are different. Filling in the proper `audio`, `text`, or `video` prefix is a lot of cognitive overhead to ask of users.

brandonocasey commented on docs/guides/audio-tracks.md:
```diff
@@ -1,69 +1,80 @@
-This is the main interface into the audio tracks of the player.
-It returns an AudioTrackList which is an array like object that contains all the `AudioTrack` on the player.
+### Player Methods
+#### `player.audioTracks()`
```
This seems more like API documentation to me, perhaps we should just make this a TODO: that links there

misteroneill replied:
Yeah, I agree. We ought to link to the API docs in this case.

brandonocasey commented on docs/guides/audio-tracks.md:
````diff
@@ -1,69 +1,80 @@
 ```js
-// get the current players AudioTrackList object
-let tracks = player.audioTracks();
+// Get the current player's AudioTrackList object.
````
it would probably be good to showcase all the events here `addtrack`, `change`, and `removetrack`.

brandonocasey replied:
It could all be done in one example, but would probably be better in multiple examples.

brandonocasey replied:
on second thought maybe we should have a general track guide/example so that it doesn't have to be repeated in every guide.

misteroneill replied:
I don't think we should showcase everything. I think the guides are meant to be a sort of tutorial or intro to working with them and not exhaustive API documentation.

brandonocasey commented on docs/guides/audio-tracks.md:
```diff
@@ -1,69 +1,80 @@
+The valid BCP 47 code for the language of the audio track, e.g. `"en"` for English or `"es"` for Spanish. For supported language translations, please see the [Languages Folder (/lang)](https://github.com/videojs/video.js/tree/master/lang) folder located in the Video.js root.
+
+#### `enabled`
+Whether or not this track should be playing. In Video.js, we only allow one track to be enabled at a time; so, if you enable more than one the last one to be enabled will end up being the only one.
```
maybe it would be good to add a note that the spec allows more than one track to be enabled, but safari and most implementations only allow one to be enabled at a time.

brandonocasey commented on docs/guides/text-tracks.md:
```diff
@@ -1,184 +1,201 @@
-- **Chapters**: Chapter titles that are used to create navigation within the video. Typically they're in the form of a list of chapters that the viewer can click on to go to a specific chapter.
-- **Descriptions**: Text descriptions of what's happening in the video for when the video portion isn't available, because the viewer is blind, not using a screen, or driving and about to crash because they're trying to enjoy a video while driving. Descriptions are read by a screen reader or turned into a separate audio track.
-- **Metadata**: Tracks that have data meant for javascript to parse and do something with. These aren't shown to the user.
+Text tracks are a feature of HTML5 video for displaying time-triggered text to the viewer. Video.js offers a cross-browser implementation of audio tracks.
```
~~audio tracks~~ text tracks

brandonocasey commented on docs/guides/text-tracks.md:
```diff
@@ -1,184 +1,201 @@
-## Adding to Video.js
-Once you have your WebVTT file created, you can add it to Video.js using the track tag. Put your track tag after all the source elements, and before any fallback content.
+## Adding Text Tracks to Video.js
+Once you have your WebVTT file created, you can add it to Video.js using the `<track>` tag. Put your `<track>` tag(s) after all the source elements and before any fallback content.
```
your WebVTT ~~file~~ files created,

brandonocasey replied:
are tracks required to be before fallback content or is that just recommended?

gkatsev replied:
There isn't any specific ordering required. It just looks nicer to group all tracks after all source elements.

brandonocasey commented on docs/guides/text-tracks.md:
```diff
@@ -1,184 +1,201 @@
+Timed text requires a text file in [WebVTT](http://dev.w3.org/html5/webvtt/) format. This format defines a list of "cues" that have a start time, an end time, and text to display. [Microsoft has a builder](https://dev.modern.ie/testdrive/demos/captionmaker/) that can help you get started on the file.
 
-When creating captions, there's also additional [caption formatting techniques] (http://www.theneitherworld.com/mcpoodle/SCC_TOOLS/DOCS/SCC_FORMAT.HTML#style) that would be good to use, like brackets around sound effects: [ sound effect ]. If you'd like a more in depth style guide for captioning, you can reference the [Captioning Key](http://www.dcmp.org/captioningkey/), but keep in mind not all  ...
+When creating captions, there are additional [caption formatting techniques](http://www.theneitherworld.com/mcpoodle/SCC_TOOLS/DOCS/SCC_FORMAT.HTML#style) to make captions more meaningful, like brackets around sound effects (e.g. `[ birds chirping ]`). For a more in depth style guide for captioning, see the [Captioning Key](http://www.dcmp.org/captioningkey/), but keep in mind not all features ar ...
```
This seems like a NOTE

brandonocasey commented on docs/guides/text-tracks.md:
```diff
@@ -1,184 +1,201 @@
+
+> **Note:** For chapters, `default` is required if you want the chapters menu to show.
+
+#### `srclang`
```
The actual property is called `language`, but it can be filled by `srclang` or `language` options as options

misteroneill replied:
According to the standard, it's `srclang` for text tracks.

misteroneill replied:
Nvm, I see what you're saying.

brandonocasey commented on docs/guides/text-tracks.md:
```diff
@@ -1,184 +1,201 @@
+#### `player.remoteTextTracks()`
+This method returns a `TextTrackList` of all the tracks that were created from `<track>` elements or that were added to the player using the `addRemoteTextTrack` method.
+
+These are distinct from the list returned by `textTracks()` because all these tracks are _removeable_, whereas tracks from `textTracks()` are not necessarily removeable.
```
This should probably just link to api docs

misteroneill replied:
Yeah, I'm simplifying this section a lot.

brandonocasey commented on docs/guides/text-tracks.md:
```diff
@@ -1,184 +1,201 @@
+
+These are distinct from the list returned by `textTracks()` because all these tracks are _removeable_, whereas tracks from `textTracks()` are not necessarily removeable.
+
+#### `player.remoteTextTrackEls()`
```
it may be good to specify somewhere what a "remote" text track means.

misteroneill replied:
It's important, I'm putting a section at the top!

brandonocasey commented on docs/guides/text-tracks.md:
```diff
@@ -1,184 +1,201 @@
+
+- `kind`
+- `label`
+- `language` (also `srclang`)
```
maybe specify that it is an alias for language

brandonocasey commented on docs/guides/tracks.md:
```diff
@@ -1,6 +1,15 @@
-* [AudioTracks](./audio-tracks.md) - allows the selection of alternative AudioTracks for a video
-* [VideoTracks](./video-tracks.md) - allows the selection of an alternative VideoTrack for a video
-* [TextTracks](./text-tracks.md) - Text Tracks are used to display subtitles and captions, and add a menu for navigating between chapters in a video.
+There are currently three types of tracks:
```
this would be a good place to put the general track examples (such as addTrack, removeTrack, getTrackById and all the track events)

misteroneill replied:
As mentioned above, I think we want to steer away from trying to do that sort of generic example.

brandonocasey commented on docs/guides/text-tracks.md:
```diff
@@ -1,184 +1,201 @@
-  * For native text tracks, this behavior depends on the browser. Some browsers will let you have multiple text tracks but others will disable all other tracks when a new one is selected.
+In general, `"descriptions"` tracks are of lower precedence than `"captions"` and `"subtitles"`. What this mean for developers using Video.js?
+
+- If you are using the `default` attribute, Video.js will choose the first track that is marked as `default` and turn it on. If there are multiple tracks marked `default`, it will turn on the first `"captions"` or `"subtitles"` track _before_ any `"descriptions"` tracks.
```
this all seems very confusing, maybe its just me though

gkatsev replied:
It is tricky. If you can think of a simpler way to say it, please let me know.

brandonocasey commented on docs/guides/video-tracks.md:
```diff
@@ -1,70 +1,83 @@
 # Video Tracks
```
It may be good to add that this is an experimental feature as there is no current support in any browser or tool for alternative video tracks.

brandonocasey reviewed (changes requested):
* It would be good to merge all the track examples into the general tracks.md file
* It would be good to link to specs and API docs for each track.

gkatsev commented on docs/guides/text-tracks.md:
```diff
@@ -1,184 +1,201 @@
-
-Text Tracks are a function of HTML5 video for providing time triggered text to the viewer. Video.js makes tracks work across all browsers. There are currently five types of tracks:
-
-- **Subtitles**: Translations of the dialogue in the video for when audio is available but not understood. Subtitles are shown over the video.
```
It looks like this section of text track types has been lost.

gkatsev replied:
Oh, I see got moved down.

gkatsev commented on docs/guides/text-tracks.md:
````diff
@@ -1,184 +1,201 @@
 ```
-One thing to be aware of is that in this case the video files themselves will *also* needs CORS headers applied to it. This is because some browsers apply the crossorigin attribute to the video source itself and not just the tracks and is considered a [security concern by the spec](https://html.spec.whatwg.org/multipage/embedded-content.html#security-and-privacy-considerations).
-
-## Track Attributes
````
The idea of this section was to showcase the track element and what you need to know about it.

gkatsev replied:
So that a user can see what should go on a track element easily without looking at the "API" section because they might think you're required to use the API.

misteroneill committed:
Address PR concerns

misteroneill commented:
I think it's covered in other comments, but I believe both @gkatsev and I agree that we shouldn't try to show generic tracks examples because the APIs are virtually identical in usage, but the actual names of things are not identical. It could cause confusion. Otherwise, I think I've addressed most concerns.
